bundles/org.openhab.binding.ambientweather/src/main/java/org/openhab/binding/ambientweather/internal/handler/AmbientWeatherEventListener.java

```diff
@@ -0,0 +1,351 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.ambientweather.internal.handler;
+
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.commons.lang.StringUtils;
+import org.json.JSONObject;
+import org.openhab.binding.ambientweather.internal.model.DevicesJson;
+import org.openhab.binding.ambientweather.internal.model.EventDataGenericJson;
+import org.openhab.binding.ambientweather.internal.model.EventSubscribedJson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+import io.socket.client.IO;
+import io.socket.client.Socket;
+import io.socket.emitter.Emitter;
+
+/**
+ * The {@link AmbientWeatherEventListener} is responsible for establishing
+ * a socket.io connection with ambientweather.net, subscribing/unsubscribing
+ * to data events, receiving data events through the real-time socket.io API,
+ * and for routing the data events to a weather station thing handler for processing.
+ *
+ * @author Mark Hilbush - Initial contribution
+ */
+public class AmbientWeatherEventListener {
+    private final Logger logger = LoggerFactory.getLogger(AmbientWeatherEventListener.class);
+
+    private static String realtimeUrl = "https://api.ambientweather.net/?api=1&applicationKey=%APPKEY%";
+
+    // Maintain mapping of handler and weather station MAC address
+    private final ConcurrentHashMap<AmbientWeatherStationHandler, String> handlers = new ConcurrentHashMap<>();
+
+    // URL used to subscribe or unsubscribe from weather data events
+    private final String subUnsubJson = "{ apiKeys: [ '%APIKEY%' ] }";
+
+    private String apiKey;
+
+    private String applicationKey;
+
+    // Socket.io socket used to access Ambient Weather real-time API
+    private Socket socket;
```

Comment:
Not so cool actually. We have a websocket factory and a http factory in the framework, not very complicated to use. Is there a reason for socket.io here?

Reply:
As mentioned in my comment below, I'm using the Jetty websocket client in another binding I'm working on, and it's working just fine. I spent quite a few hours trying to get the Jetty websocket client to work with the Ambient Weather backend, but was unable to make it work (was unable to make a successful connection with the backend service). That was the reason for switching to socklet.io. For the socket.io libs, I tried to stay in sync with the ones openhabcloud is using.